src/ameba/ast/variabling/variable.cr

```diff
@@ -137,10 +137,10 @@ module Ameba::AST
     # `false` otherwise.
     def target_of?(assign)
       case assign
-      when Crystal::Assign           then eql?(assign.target)
-      when Crystal::OpAssign         then eql?(assign.target)
-      when Crystal::MultiAssign      then assign.targets.any? { |target| eql?(target) }
-      when Crystal::UninitializedVar then eql?(assign.var)
+      when Crystal::UninitializedVar          then eql?(assign.var)
+      when Crystal::Assign, Crystal::OpAssign then eql?(assign.target)
+      when Crystal::MultiAssign
+        assign.targets.any? { |target| eql?(target) }
```

Comment:
I'm not sure if this change (and similar above) is a good choice.

This aspect has not yet been mentioned in the discussion or anywhere in the documentation and specs: some conditions have an effect on flow typing and merging branches affects typing.

Here the different branches have technically the same body. But combining them introduce a union type `Crystal::Assign | Crystal::OpAssign` which can lead to all kinds of potentially unintended consequences.

If we keep the branches separate, the type of `assign` is restricted to the respective type.

The compiler uses this pattern intentionally in a number of places (including ones dealing with switching over AST nodes). So I would assume it's also intentional here.

Reply:
I don't see how this affects dispatch, see here: https://carc.in/#/r/hqb4